packages/flutter/lib/src/widgets/shape_boundary.dart

```diff
@@ -0,0 +1,142 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+
+import 'framework.dart';
+
+/// An abstract class that defines the boundaries of a geometric shape.
+///
+/// [isWithinBoundary] Returns whether the specified shape is within the boundary
+///
+/// [nearestShapeWithinBoundary] Returns the shape closest to the specified shape
+/// within the boundary, or returns the shape as is if it is within the boundary.
+///
+/// See also:
+/// * [OffsetBoundaryProvider], A widget that provides the [Offset] of a [ShapeBoundary]
+/// * [RectBoundaryProvider], A widget that provides the [Rect] of a [ShapeBoundary]
```

Comment:
nit: maybe this should say `A widget that provides a [ShapeBoundary] for a [Rect].`